File: 15-3sum/15-3sum.scala
import scala.collection.mutable._

object Solution {
    def threeSum(nums: Array[Int]): List[List[Int]] = {
        var res: List[List[Int]] = List()
        
        val sorted = nums.sorted(Ordering.Int)
        
        for(i <- 0 to sorted.length - 3) {
            if (i == 0 || sorted(i) != sorted(i - 1)) {
                val pivot = sorted(i)
                var s = i + 1
                var e = sorted.length - 1

                while(s < e) {
                    val sum = sorted(s) + sorted(e) + pivot

                    if (sum == 0) {
                        res = res :+ List(sorted(s), sorted(e), pivot)
                        var ns = s + 1
                        var ne = e - 1
                        while(ns < sorted.length && sorted(ns) == sorted(s)) ns += 1
                        while(ne >= 0 && sorted(ne) == sorted(e)) ne -= 1
                        s = ns
                        e = ne
                    } else if (sum < 0) {
                        s += 1
                    } else {
                        e -= 1
                    }
                }
            }
        }
        
        res
    }
}

File: 15-3sum/15-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Set<Integer> dup = new HashSet<>();
        
        for(int i = 0; i < nums.length - 2; i++) {
            Set<Integer> set = new HashSet<>();
            for(int j = i + 1; j < nums.length; j++) {
                int target = -(nums[i] + nums[j]);
                if (set.contains(target)) {
                    List<Integer> cand = List.of(nums[i], nums[j], target);
                    int hash = hashCode(cand);
                    if (!dup.contains(hash)) {
                        dup.add(hash);
                        res.add(cand);
                    }
                }
                set.add(nums[j]);
            }
        }
        
        return res;
    }
    
    int hashCode(List<Integer> list) {
        int[] arr = new int[] { list.get(0), list.get(1), list.get(2) };
        Arrays.sort(arr);
        return (arr[0] + ";" + arr[1] + ";" + arr[2]).hashCode();
    }
}